### ISSUE TITLE
Decision Tree process implementation design

### ISSUE THREAD

**Issue by john0x44** [john0x44/Censorship-Detection-In-Censored-Planet#97]
Currently doing decision tree implementation process in order to generate decision trees for each domain to analyse censorship patterins using the RES matrix (CenDTect: Equation 3) which is implemented now from the DataAggregator class. The decision trees will be trained with ScikitLearn DecisionTreeClassifier library using Gini Impurity and then classify IP organizations as showing Likely Censorship, then move onto clustering of domains with similar censorship behaviors (CenDTect : Section IV). Building on the prior issues (#93 and #95) which produced varied t_vector distribution in weights, we can outlint the process of the decision tree implementation 

Each tree uses the RES matrix (rows will have IP organizations and columns : with time and response classes) to predict a value (1 for Likely Censorship or 0 otherwise) we utilize Gini Impurity in order to find proper tree split decision to seperate censored from non-censored IP organizations (CenDTect : Equation 3). Then with this decision tree implementation it enables cross classification distances (Equation 4) for clustering domains with simultaenous blocking behaviors (CenDTect: Section IV)

Implementation process and research : 

The input preparation will use the RES matrix generated from the DataAggregator class with rows for IP organizations and columns for domain, IP Org, country and time (t1 to tm) Each time column has a t_vector that has the 8 class responses we implemented on issue #96, and contains the appropriate distributed weights via the oversampling method. 

For each domain we take the t_vector and split into seperate columns, then group by IP Org we then combine rows that have the same IP organization. On (CenDTect: Section IV-A, Equation 3) the decision tree needs a table where each row represents an IP organization and columns showing how response classes change over time, which helps the tree figure out if an IP organization is blocking a website. 

Deciding If censored or not : 

The decision trees needs to know what to predict, so we create labels for each IP organization. Meaning for each IP organization we check the t_vector lists across the time, and focus only on response classes that provide a (Likely Censorship scenario in which (CenDTect looks at :"For this paper, we conservatively only report the clusters of the “Likely Censorship” type, which signifies clear blocking signals, and helps our model reflect censorship events accurately") 

So if the sum of the weights on response classes is more than a set threshold we set the label to a likely censorship, otherwise not. The paper notes that the average depth of their trees are around 25 so we need to log the depths as well if the average depth is not around 25 then we need to adjust some variation. 

Documentation reading : 

https://scikit-learn.org/stable/modules/tree.html
https://www.geeksforgeeks.org/machine-learning/decision-tree-implementation-python/#
https://www.analyticsvidhya.com/blog/2021/08/decision-tree-algorithm/
